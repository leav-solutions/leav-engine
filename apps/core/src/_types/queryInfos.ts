// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbProfiler} from './dbProfiler';
import {IValueVersion} from './value';

export interface IQueryInfos {
    userId?: string;
    lang?: string;
    queryId?: string;
    version?: IValueVersion;
    treeId?: string;
    groupsId?: string[];
    applicationId?: string;
    dbProfiler?: IDbProfiler;
}
