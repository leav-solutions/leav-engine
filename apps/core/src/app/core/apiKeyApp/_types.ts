// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IApiKeyFilterOptions} from '_types/apiKey';
import {IPaginationParams, SortOrder} from '_types/list';

export interface IApiKeysArgs {
    filters?: IApiKeyFilterOptions;
    pagination: IPaginationParams;
    sort: {
        field: 'id';
        order: SortOrder;
    };
}

export interface ISaveApiKeyArgs {
    apiKey: {
        id: string;
        label: string;
        expiresAt: number;
        userId: string;
    };
}

export interface IDeleteApiKeyArgs {
    id: string;
}
