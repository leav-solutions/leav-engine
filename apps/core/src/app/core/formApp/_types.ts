// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IForm, IFormDependentElements, IFormElement} from '_types/forms';
import {IPaginationParams, ISortParams} from '_types/list';

export type IFormForGraphql = Omit<IForm, 'elements'> & {elements: IFormDependentElementsForGraphQL[]};
export type IFormDependentElementsForGraphQL = Omit<IFormDependentElements, 'elements'> & {
    elements: IFormElementForGraphQL[];
};
export type IFormElementForGraphQL = Omit<IFormElement, 'settings'> & {settings: Array<{key: string; value: any}>};

export interface IGetFormArgs {
    filters: ICoreEntityFilterOptions & {library: string; system?: boolean};
    pagination: IPaginationParams;
    sort: ISortParams;
}

export interface IGetRecordFormArgs {
    recordId: string;
    libraryId: string;
    formId: string;
}

export interface ISaveFormArgs {
    form: IFormForGraphql;
}

export interface IDeleteFormArgs {
    library: string;
    id: string;
}
