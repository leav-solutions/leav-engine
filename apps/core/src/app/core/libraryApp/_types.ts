// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {LibraryBehavior} from '_types/library';
import {SortOrder} from '_types/list';

export interface IGetLibraryParams {
    filters?: {id?: string[]; label?: string[]; system?: boolean; behavior?: LibraryBehavior[]};
    pagination?: {limit: number; offset: number};
    sort?: {field: string; order: SortOrder};
    strictFilters?: boolean;
}
