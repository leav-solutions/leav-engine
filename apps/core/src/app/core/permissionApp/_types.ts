// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IPermissionsTreeTarget, PermissionsActions, PermissionTypes} from '_types/permissions';

export interface IInheritedPermissionsQueryParams {
    type: PermissionTypes;
    applyTo: string;
    actions: PermissionsActions[];
    userGroupNodeId: string;
    permissionTreeTarget: IPermissionsTreeTarget;
}
