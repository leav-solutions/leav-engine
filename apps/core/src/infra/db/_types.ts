// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
