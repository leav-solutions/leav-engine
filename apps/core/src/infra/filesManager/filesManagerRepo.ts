// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IQueryInfos} from '_types/queryInfos';
import {IRecord} from '_types/record';
import {aql, join} from 'arangojs/aql';
import {IDbDocument} from 'infra/db/_types';
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {join as joinPath} from 'path';
import winston from 'winston';
import {FilesAttributes} from '../../_types/filesManager';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.utils.logger'?: winston.Winston;
}

export interface IFilesManagerRepo {
    getRecord(
        {fileName, filePath, fileInode}: {fileName: string; filePath: string; fileInode?: number},
        {recordLibrary, recordId}: {recordLibrary: string; recordId?: string},
        retrieveInactive: boolean,
        ctx: IQueryInfos
    ): Promise<IRecord | null>;
    getParentRecord(fullParentPath: string, library: string, ctx: IQueryInfos): Promise<IRecord | null>;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.utils.logger': logger = null
}: IDeps): IFilesManagerRepo {
    return {
        async getRecord(
            {fileName, filePath, fileInode}: {fileName: string; filePath: string; fileInode?: number},
            {recordLibrary, recordId}: {recordLibrary: string; recordId?: string},
            retrieveInactive: boolean,
            ctx: IQueryInfos
        ): Promise<IRecord | null> {
            const coll = dbService.db.collection(recordLibrary);

            const queryParts = [aql`FOR r in ${coll}`];
            let results: IDbDocument[];
            if (recordId) {
                queryParts.push(aql`FILTER (r.id == ${recordId})`);
            } else if (!!fileInode) {
                queryParts.push(aql`FILTER (r.${FilesAttributes.INODE} == ${fileInode}
                        OR (r.${FilesAttributes.FILE_PATH} == ${filePath} AND r.${FilesAttributes.FILE_NAME} == ${fileName})
                    )`);
            } else {
                queryParts.push(aql`FILTER (r.${FilesAttributes.FILE_PATH} == ${filePath}
                        AND r.${FilesAttributes.FILE_NAME} == ${fileName}
                    )`);
            }
            if (!retrieveInactive) {
                queryParts.push(aql`FILTER r.active == true`);
            }
            queryParts.push(aql`SORT r._key DESC`);
            queryParts.push(aql`RETURN MERGE(r, {library: ${recordLibrary}})`);
            const query = join(queryParts, '\n');

            try {
                results = await dbService.execute({
                    query,
                    withTotalCount: false,
                    ctx
                });
            } catch (e) {
                return null;
            }

            const count = results.length;
            if (count === 0) {
                return null;
            }

            if (count > 1) {
                logger.warn(`[FilesManager] Multiple record found using fileName and filePath: ${results.toString()}`);
            }

            return dbUtils.cleanup(results[0]);
        },
        async getParentRecord(fullParentPath: string, library: string, ctx: IQueryInfos): Promise<IRecord | null> {
            let parentPath = fullParentPath.split('/');
            const parentName = parentPath.pop();

            if (parentPath.length === 0) {
                parentPath = ['.'];
            }

            const coll = dbService.db.collection(library);
            const query = aql`FOR r in ${coll} FILTER (r.${FilesAttributes.FILE_NAME} == ${parentName} AND r.${
                FilesAttributes.FILE_PATH
            } == ${joinPath(...parentPath)}) RETURN MERGE(r, {library: ${library}})`;
            let results: IDbDocument[];
            try {
                results = await dbService.execute({
                    query,
                    withTotalCount: false,
                    ctx
                });
            } catch (e) {
                logger.warn(`[FilesManager] Error when search parent folder : ${fullParentPath}`);
                return null;
            }

            const parent = results[0] ? dbUtils.cleanup(results[0]) : null;

            return parent;
        }
    };
}
