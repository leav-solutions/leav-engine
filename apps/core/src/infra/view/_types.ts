// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
