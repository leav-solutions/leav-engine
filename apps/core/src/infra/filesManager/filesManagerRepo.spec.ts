// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IQueryInfos} from '_types/queryInfos';
import filesManagerRepo from './filesManagerRepo';

describe('FilesManagerRepo', () => {
    test('getRecord', async function () {
        const ctx: IQueryInfos = {
            userId: '0',
            queryId: 'filesManagerRepoTest'
        };
        const mockEnsureIndex = jest.fn();
        const mockCollection = new Database().collection('test');
        mockCollection.ensureIndex = mockEnsureIndex;

        const mockDb = new Database();
        mockDb.collection = jest.fn().mockReturnValue(mockCollection);

        const mockDbServ = {
            db: mockDb,
            execute: global.__mockPromise([])
        };

        const repo = filesManagerRepo({
            'core.infra.db.dbService': mockDbServ
        });

        const res = await repo.getRecord(
            {
                fileName: 'fileNameTest',
                filePath: 'filePathTest',
                fileInode: 128
            },
            {
                recordLibrary: 'test',
                recordId: '128'
            },
            false,
            ctx
        );
        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
        expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
    });
    test('getParentRecord', async function () {
        const ctx: IQueryInfos = {
            userId: '0',
            queryId: 'filesManagerRepoTest'
        };
        const mockEnsureIndex = jest.fn();
        const mockCollection = new Database().collection('test');
        mockCollection.ensureIndex = mockEnsureIndex;

        const mockDb = new Database();
        mockDb.collection = jest.fn().mockReturnValue(mockCollection);

        const mockDbServ = {
            db: mockDb,
            execute: global.__mockPromise([])
        };

        const repo = filesManagerRepo({
            'core.infra.db.dbService': mockDbServ
        });

        const res = await repo.getParentRecord('fullParentPath', 'libraryTest', ctx);
        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
        expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();
    });
});
