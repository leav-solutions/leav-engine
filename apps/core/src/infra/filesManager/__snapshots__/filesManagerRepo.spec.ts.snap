// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilesManagerRepo getParentRecord 1`] = `"FOR r in @@value0 FILTER (r.@value1 == @value2 AND r.@value3 == @value4) RETURN MERGE(r, {library: @value5})"`;

exports[`FilesManagerRepo getParentRecord 2`] = `
{
  "@value0": "test",
  "value1": "file_name",
  "value2": "fullParentPath",
  "value3": "file_path",
  "value4": ".",
  "value5": "libraryTest",
}
`;

exports[`FilesManagerRepo getRecord 1`] = `
"FOR r in @@value0
FILTER (r.id == @value1)
FILTER r.active == true
SORT r._key DESC
RETURN MERGE(r, {library: @value2})"
`;

exports[`FilesManagerRepo getRecord 2`] = `
{
  "@value0": "test",
  "value1": "128",
  "value2": "test",
}
`;
